torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,863 @@
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+)
+from weakref import ref as WeakRef
+
+import torch.fx
+from torch import Tensor
+
+from torch._C import _cuda_CUDAAllocator_AllocatorState as AllocatorState
+from torch._dynamo.mutation_guard import GenerationTracker
+
+from torch._inductor.compile_fx import (
+    get_expanded_dims,
+    index_expanded_dims,
+    remove_unaligned_input_idxs,
+    static_input,
+)
+from torch._prims_common import check
+
+from torch.utils import _pytree as pytree
+from . import config
+
+
+@dataclasses.dataclass(frozen=True)
+class GraphID:
+    "Unique counter of a cuda graph recording"
+    id: int
+
+
+@dataclasses.dataclass(frozen=True)
+class FunctionID:
+    "Unique counter of a function wrapped in cudagraphify_impl"
+    id: int
+
+
+@dataclasses.dataclass(frozen=True)
+class WrappedFunction(object):
+    model: Callable
+    static_input_idxs: Sequence[int]
+    id: FunctionID
+
+
+class TreeManagerContainer(object):
+    """
+    Manages the lifetime of the tree manager. Like `PrivatePool` in cuda caching allocator,
+    the tree and its corresponding memory pool should be kept alive as long as any outstanding
+    graph or tensor which is an output of a graph remains alive.
+    """
+
+    def __init__(self):
+        # This class keeps a strong reference to tree_manager,
+        # but upon all other strong references to the tree_manager will reset it to None.
+        # We need a strong reference so that we can still access its attributes upon cleanup.
+        self.tree_manager: Optional[CUDAGraphTreeManager] = None
+
+        # Number of outstanding references to the current tree manager
+        self.live_cudagraphify_fns = 0
+
+        # Following two objects are only set in the case that Tensor outputs outlive
+        # the cudagraphify_fns. Reference to the Graph is needed to keep the private pool from
+        # deallocation.
+        self.live_tensors_count = 0
+        self.graph: Optional[torch.cuda.CUDAGraph] = None
+
+        self.lock = threading.Lock()
+
+    def finalize_tensor(self):
+        with self.lock:
+            self.live_tensors_count -= 1
+            if self.live_tensors_count == 0:
+                self.graph = None
+
+                # manager was used again after existing cleanup,
+                # we shouldnt set it to None
+                if self.live_cudagraphify_fns == 0:
+                    self.tree_manager = None
+
+    def finalize_reference(self):
+        with self.lock:
+            self.live_cudagraphify_fns -= 1
+            if self.live_cudagraphify_fns == 0:
+                self.finalize_tree_manager()
+
+    def add_strong_reference(self, fn: Callable):
+        with self.lock:
+            self.live_cudagraphify_fns += 1
+
+        weakref.finalize(fn, self.finalize_reference)
+
+    def finalize_tree_manager(self):
```

Comment:
It would be extremely helpful if methods like this which are never called externally have a leading underscore, showing they're class private.